Jermolene/TiddlyWiki5
# Translateable plugins - first try

(This info looks much more complicated than it actually is)

If you download this branch you can do the following things: 

# Pluginlibrary Server - NEW

- **Edit 2022-04-03**: This is a new PR now: **Pluginlibrary local test edition** #6586 
  - go to `./editions/pluginlinlibrary-server` ... new directory ... CAN be a separated PR (@Jermolene ??)
    - call `npm test` ... this will create a local library version and run a server on http://localhost:8080
    - It will allow you to test load plugins to the server
    - There is a "Delete all "plugin modules"", which will remove all plugins.
      - see: http://localhost:8080/#Remove-all-plugin-modules
    - call `npm start` ... if you want to skip the "build library" step

# Switch language for plugins

 - You need to install the TW core "menubar" plugin  to test this branch
 - You need to install the TW core "de-DE" language to test this branch
 - This PR contains a POC translation to **de-DE**. 
   - see file: `menubar.multids`
   - ~~Only 2 tiddlers are translated atm.~~ 
   - Contents (caption) -> Inhalt ... So it's easy to test and see if it works, since the Contents text is seen by default top left.
   - Contents (description) -> This text is part of the **plugin : config** tab
     - more commits will follow to translate everything

# Structure

## initial language

 - The initial language is part of the plugin
   - That's important, so there always will be a "fallback" text, if any text translation is missing
   - Technically is a "language" 
 - I did use the directory name `_en-GB` the underscore guarantees, that the directory is listed first place. 
    - The directory name doesn't matter. 
    - It's only the "namespace", "author" and "plugin-name" that matters: eg:
      - `title: $:/language/tiddlywiki/menubar` 
      - `title: $:/language/<author>/<plugin-name>`
    - Important: `/<author>/<plugin-name>` are used for AddOns too!

## plugin: `plugin.info`

There is a new parameter, which indicates, that a plugin is "ready to be translated". 

  - `addons": ["language-addon"]` .. it's an array. ATM only `language-addon` is supported!
     -  as Jeremy suggested, it may be possible to create `theme-addon` plugins, **not part of this PR!**
  - This parameter is part of the plugin: "plugin-info"
  - This parameter is used in the add-plugin-UI to show a new button.

## language: `plugin-info`

The translation `plugin-type` must be `language-addon` and is part of the `languages` namespace. see:

{
	"title": "$:/languages/de-DE/tiddlywiki/menubar",
	"name": "Menü Toolbar/de-DE",
	"plugin-type": "language-addon",
	"description": "Top Menü Plugin",
	"author": "MarioPietsch aka pmario",
	"core-version": ">=5.1.24"
}

 - The **title** and **plugin-type** are important to identify the addon
 - The language plugin title is the ONLY tiddler, that contains the "language string" eg: **de-DE**
 - the `title` must have the structure: 
   - `$:/languages/de-DE/tiddlywiki/menubar`
   - `<language-string>/<author>/<plugin-name>`
     - the `<language-string>` is used by the language-switcher UI and is stored in [$:/language](http://localhost:8080/#%24%3A%2Flanguage) tiddler
     - It's used to identify the "root" language namespace. 
     - Everything after the `<language-string>` is part of the "addon-namespace"

## Video

https://youtu.be/yZMaZ4qeaKk
